### DISCUSSION TITLE
Intercepting Copy and Paste

### DISCUSSION THREAD

**Discussion by amkraft** [tauri-apps/tauri#8446] [Q&A]
for reference, I develop on a Mac, I intend to build to Mac/Windows/Linux.

This question follows me having read about the requirement for Copy Paste system menu items to enable the behavior in app.

In summary, I think what I'm looking for is the ability to intercept the system level copy/paste behavior to allow or disallow it based on my own criteria. Is there a way to replace this:

```rust
.add_native_item(MenuItem::Copy)
```

with

```rust
.add_item(CustomMenuItem::new("copy".to_string(), "Copy"))
// with an on_menu_event
"copy" => {
  // here is some if/else criteria,
  run_system_copy(); // here, somehow call the system level copy behavior
}

// same goes for "paste"
```

Alternatively, is there a way to **enable system level Copy/Paste *without* adding those menu items**?

Specifically,

I have an app that is basically a graphic design tool (Adobe Illustrator etc...). I have `<input type="text">` and `<textarea>` scattered throughout, so, I really need that sytem level copy/paste behavior, I think it would be too difficult to re-implement this from scratch. But also, I have a graphical canvas with shapes, selection, and the ability for the user to copy/paste shapes.

As a browser app (no menu items, only keyboard), my app works fine. I use a keyboard handler to check which element is active (form type or canvas), and proceed accordingly. (preventDefault or not)

As a Tauri app with a menu with native_items Copy/Paste, the same keyboard handler continues to work fine. Except, obviously, trying to copy/paste the canvas shapes by using the System Menu items (not using keyboard shortcuts). Hence, my situation:

- if those menu items were gone, the UX would make sense (enough).
- if the menu items are present, I need them to be able to fork to the default behavior or custom behavior, depending on my criteria.

Thanks for reading through my problem.

**Comment by FabianLars** [comment]
Off topic but i'm interesting in that part

> As a browser app (no menu items, only keyboard), my app works fine. I use a keyboard handler to check which element is active (form type or canvas), and proceed accordingly. (preventDefault or not)

How does your browser app not have menu items? I thought browsers themselves set them the same way tauri does (or similar enough).
I'm fairly new to macos so i'm genuinely curious.

**Reply by amkraft** [reply]
ah I could have said that better. "as a browser app", meaning, if I don't use Tauri. If I simply build a Javascript web app. In which case there is a menu, yes, Firefox's menu, which has nothing to do with my app, so I just ignore it. This is on MacOS at least.

Is that what meant?

**Comment by amkraft** [comment]
tldr, my original request:

- can I intercept/customize/extend the System menu "copy/paste" on click events? or,
- can I enable copy/paste keyboard shortcuts without including the system menu "copy/paste" menu items? or make those system menu items invisible?

alternatively, if anyone wants to help, can you point me towards the code in Tauri that handles this system menu copy paste behavior? maybe I can look into it myself.

Thanks all!
